homeassistant/components/epic_games_store/calendar.py

```diff
@@ -0,0 +1,110 @@
+"""Calendar platform for a Epic Games Store."""
+
+from __future__ import annotations
+
+from collections import namedtuple
+from datetime import datetime
+from typing import Any
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DOMAIN
+from .coordinator import CalendarType, EGSCalendarUpdateCoordinator
+
+DateRange = namedtuple("DateRange", ["start", "end"])
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the local calendar platform."""
+    coordinator: EGSCalendarUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+
+    entities = [
+        EGSCalendar(coordinator, entry.entry_id, CalendarType.FREE),
+        EGSCalendar(coordinator, entry.entry_id, CalendarType.DISCOUNT),
+    ]
+    async_add_entities(entities, True)
+
+
+class EGSCalendar(CoordinatorEntity[EGSCalendarUpdateCoordinator], CalendarEntity):
+    """A calendar entity by Epic Games Store."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        coordinator: EGSCalendarUpdateCoordinator,
+        config_entry_id: str,
+        cal_type: CalendarType,
+    ) -> None:
+        """Initialize EGSCalendar."""
+        super().__init__(coordinator)
+        self._cal_type = cal_type
+        self._attr_translation_key = f"{cal_type}_games"
+        self._attr_unique_id = f"{config_entry_id}-{cal_type}"
+        self._attr_device_info = DeviceInfo(
+            entry_type=DeviceEntryType.SERVICE,
+            identifiers={(DOMAIN, config_entry_id)},
+            manufacturer="Epic Games Store",
+            name="Epic Games Store",
+        )
+        self._event: CalendarEvent | None = None
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+        return self._event
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return the state attributes."""
+        return {
+            "games": self.coordinator.data[self._cal_type],
```

Comment:
What does this return? We want to avoid putting a lot of data in these state attributes as they can blow up the state machine

Reply:
`TypeError` or `KeyError`, more a Type as on test `tests/components/epic_games_store/test_config_flow.py::test_form_cannot_connect_wrong_param`

See

```shell
2023-12-28 11:01:23.456 WARNING  MainThread homeassistant.components.epic_games_store.config_flow:config_flow.py:48 [{'message': 'CatalogQuery/searchStore: Request failed with status code 400', 'locations': [{'line': 18, 'column': 9}], 'correlationId': 'e10ad58e-a4f9-4097-af5d-cafdbe0d8bbd', 'serviceResponse': '{"errorCode":"errors.com.epicgames.catalog.invalid_country_code","errorMessage":"Sorry the value you entered: en-US, does not appear to be a valid ISO country code.","messageVars":["en-US"],"numericErrorCode":5222,"originatingService":"com.epicgames.catalog.public","intent":"prod","errorStatus":400}', 'stack': None, 'path': ['Catalog', 'searchStore']}]
2023-12-28 11:01:23.456 ERROR    MainThread homeassistant.components.epic_games_store.config_flow:config_flow.py:82 Unexpected exception
Traceback (most recent call last):
  File "/workspaces/home-assistant-core/homeassistant/components/epic_games_store/config_flow.py", line 78, in async_step_user
    await validate_input(self.hass, user_input)
  File "/workspaces/home-assistant-core/homeassistant/components/epic_games_store/config_flow.py", line 51, in validate_input
    data["data"]["Catalog"]["searchStore"]["elements"]
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

But this case should not happen now as of new version of lib, so I refactored it in c0a3429e6063c38f7f67d0d56159df7f5077b2fd

Reply:
Yes we can! Since this PR can't make 2024.2, I think we can make some nice services in a follow up to have it ready with 2024.3

Reply:
Fixed in 73c20f34803b0a0ec242bf0740494f17a68f6f59

Reply:
Fixed in f3d475385aa2fb8e6c994237e000270e4ab095f8